<div class="alert alert-block alert-success">
<b>Comentario del Revisor Iteración 2.</b>

# Comentario General #2

Hola, Pablo.  

Me presento: soy David Bautista, el revisor humano que te acompañará en el proceso de revisión del proyecto después de la ayuda inicial de Dot.  
En pocas palabras, quiero felicitarte por las modificaciones realizadas; he dejado algunos comentarios como retroalimentación sobre los puntos que habia que reforzar tras la primera iteración.  
Asimismo, te felicito por la culminación del proyecto y te deseo mucho ánimo en la continuación de este proceso de aprendizaje.  
</div>



<div style="border:solid green 2px; padding: 20px">

👋 **¡Hola! Soy Dot, tu revisor de IA.**

He completado la primera revisión de tu código. A continuación, encontrarás mis comentarios y sugerencias de mejora.

**¿Qué debes hacer ahora?**

1. **Revisar:** Lee mis comentarios en el notebook más abajo.
2. **Decidir:**
* **¿Estás de acuerdo con el feedback?** ¡Genial! Actualiza tu código según las sugerencias.
* **¿No estás de acuerdo o crees que es un error?** ¡No hay problema! Puedes dejar tu código exactamente como está.


3. **Volver a enviar:** Envía tu proyecto de nuevo. **Tu próxima iteración será revisada por un experto humano**, independientemente de si hiciste cambios o no.

-------------------------------------------------------------------------------------------------------------
Mis comentarios están más abajo. **Te pido amablemente que no los muevas, modifiques ni elimines**.

Verás mis comentarios resaltados en cuadros verdes, amarillos o rojos como estos:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correcto. Todo se ha hecho exitosamente.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones.
</div>



# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
import pandas as pd

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Pandas se importa correctamente con un alias claro (`pd`), lo cual coincide con la convención utilizada en todo el proyecto.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
filepath='/datasets/music_project_en.csv'
df = pd.read_csv(filepath)

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

El dataset se carga correctamente en `df` usando `pd.read_csv()`, y definir una variable `filepath` mejora la legibilidad y facilita la reutilización.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [3]:
df.info()
df.columns


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. El índice `  UserID` tiene dos espacios en blanco antes, no está escrito como snake_case y `  City  ` tiene dos espacios antes y después del nombre.




<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

En tus notas sobre el formato de los encabezados, resulta un poco confuso porque los ejemplos de nombres de columnas con espacios extra no coinciden exactamente con los nombres de columnas reales que se muestran más adelante (p. ej., espacios al inicio/al final). Considera hacer referencia a los nombres de columnas exactamente tal como aparecen (copiar/pegar desde `df.columns`) para que no haya ambigüedad sobre qué columnas necesitan limpieza.<br>
</div>

### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Se notó que todas las columnas arrojan `object` al realizar el método `.dtypes`. Para indagar un poco más en que tipos de datos son se tomó la libertad de realizar el siguiente código de chequeo:
    ```
    datos = [df[i][7] for i in df]
    tipos = [type(df[i][7])for i in df]
    print(tipos,'\n', datos)
    ```
    Este código muestra la comparación de valores de la fila 8 con su tipo de dato. Lo que muestra es lo siguiente:

    ```
    [<class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>] 
    ['F03E1C1F', 'Feeling This Way', 'Polina Griffith', 'dance', 'Springfield', '20:47:49', 'Wednesday']
    ```
    Esto indica que todos los elementos de las filas son de tipo string. Así que, podría ser que para la finalidad del proyecto haya que cambiar algunas columnas al tipo correcto.

2.  Finalidad del proyecto: "Analizar si la actividad de los usuarios y las usuarias presenta variaciones según el día de la semana y la ciudad."

    El dataframe cuenta con la información necesaria como el día de la semana (`day`), la hora (`time`) y la ciudad (`city`) además de la información de las canciones como el nombre de la canción (`track`), el artista (`artist`) y el género (`genre`). Esta información es suficiente para realizar estudios analíticos y llegar a conclusiones sólidas.


    El análisis menciona usuarios y usuarias; sin embargo, no existe en el DataFrame una columna que distinga el sexo de cada usuario por lo que se tendrá que englobar usuarios y usuarias en una misma categoría.

3. Para revisar valores tipo NaN se utilizó la función `info()` que nos arroja:

    ```
    <class 'pandas.core.frame.DataFrame'>
    RangeIndex: 65079 entries, 0 to 65078
    Data columns (total 7 columns):
     #   Column   Non-Null Count  Dtype 
    ---  ------   --------------  ----- 
     0   user_id  65079 non-null  object
     1   track    63736 non-null  object
     2   artist   57512 non-null  object
     3   genre    63881 non-null  object
     4   city     65079 non-null  object
     5   time     65079 non-null  object
     6   day      65079 non-null  object
    dtypes: object(7)
    memory usage: 3.5+ MB
    ```
    Se Realizó un código que muestra cada columna con un conteo de sus valores NaN.

   ```
   #crea una lista con los nombres de las columnas y una lista con los NAN totales para cada una:
    son_nan = [df[i].isna().sum() for i in df]
    columnas =[i for i in df.columns]
    #Crea una lista con strings como elementos y parea los elementos de las listas previamente creadas 
    valores_NAN = [f"La columna '{columnas}' tiene {son_nan} valores NAN" for columnas, son_nan in zip(columnas, son_nan)]
    #imprime la lista de manera estética
    print(*valores_NAN, sep='\n')
    #alternativamente se puede usar:
    df.isna().sum()
   ```
   El código resulta en:

   ```
    La columna 'user_id' tiene 0 valores NAN
    La columna 'track' tiene 1343 valores NAN
    La columna 'artist' tiene 7567 valores NAN
    La columna 'genre' tiene 1198 valores NAN
    La columna 'city' tiene 0 valores NAN
    La columna 'time' tiene 0 valores NAN
    La columna 'day' tiene 0 valores NAN

    user_id       0
    track      1343
    artist     7567
    genre      1198
    city          0
    time          0
    day           0
    dtype: int64
   ```
   Para valores duplicados basta con usar `df.duplicated().sum()` para obtener `np.int64(3826)` Lo que quiere decir que existen un total de 3826 duplicados exactos de filas en el DataFrame.

   Respecto a los tipos de datos, como se comento en la respuesta 1. Todos son de tipo string, por lo que, tal vez hay que transformar algunos de ellos antes de trabajarlos.


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Esta sección te pide responder a las preguntas 1–3, pero tus respuestas están numeradas como 1, 2 y 4. Considera añadir el elemento que falta para la pregunta 3 (aunque ya hayas mencionado valores faltantes/duplicados/tipos en otro lugar) para que el entregable coincida con la estructura del enunciado.<br>
</div>

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [4]:
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [5]:
nuevos_indices = [i.strip().lower().replace(' ', '_').replace('userid', 'user_id') for i in df.columns]
df.columns = nuevos_indices
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [10]:
#Todo lo necesario se redujo en el paso anterior (Etapa2.2)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [13]:
#Todo lo necesario se realizo en la Etapa 2.2
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Aunque ya obtuviste `user_id` anteriormente, este paso te pide específicamente que renombres `userid` a `user_id` aquí (normalmente usando una operación de renombrado dedicada). Considera mostrar ese renombrado explícito en este paso para demostrar la técnica solicitada.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [12]:
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [6]:
#Método simple
print(df.isna().sum())
print()

#Método mostrado en el análisis de Etapa 1.1:
#crea una lista con los nombres de las columnas y una lista con los NAN totales para cada una:
son_nan = [df[i].isna().sum() for i in df]
columnas =[i for i in df.columns]
#Crea una lista con strings como elementos y parea los elementos de las listas previamente creadas 
valores_NAN = [f"La columna '{columnas}' tiene {son_nan} valores NAN" for columnas, son_nan in zip(columnas, son_nan)]
#imprime la lista de manera estética
print(*valores_NAN, sep='\n')

# Método info.... Hay que hacer una resta pero es trivial, Aunque siepre es bueno comprobar
print(df.info())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

La columna 'user_id' tiene 0 valores NAN
La columna 'track' tiene 1343 valores NAN
La columna 'artist' tiene 7567 valores NAN
La columna 'genre' tiene 1198 valores NAN
La columna 'city' tiene 0 valores NAN
La columna 'time' tiene 0 valores NAN
La columna 'day' tiene 0 valores NAN
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    63736 non-null  object
 2   artist   57512 non-null  object
 3   genre    63881 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [7]:
#Crea la lista de columnas a reemplazar evitando errores humanos de captura de strings
columns_to_replace = [i for i in df.columns[1:4]]
# por si el esquema falla Use la lista de abajo en su lugar:
#columns_to_replace = ['track', 'artist', 'genre']
#print(columns_to_replace)
#Bucle que rellena los valores NAN con el string 'unknown'
for i in columns_to_replace:
    df[i].fillna('unknown', inplace=True)
#Para el Sr. DOT: por eso el print está comentado porque revisé bien que funcionara mi list comprehension, también revisé que número de columna correspondía a cada uno, esto lo usé para este ejemplo específico. Cámara DOT, no es una función.... 

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Para `columns_to_replace`, la tarea pide una lista explícita de los nombres de las columnas objetivo (`['track', 'artist', 'genre']`). Construirla mediante slicing depende del orden de las columnas y puede fallar silenciosamente si el esquema cambia, por lo que es más claro y seguro enumerar los nombres de las columnas directamente.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [8]:
# Cuenta los valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Buen paso de verificación: volviste a contar los valores faltantes después del reemplazo, confirmando que las columnas especificadas ya no contienen nulls.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [20]:
# Cuenta los duplicados explícitos
# Muestra el total de filas exactamente iguales
print(df.duplicated().sum())
# Muestra el total de elementos repetidos en una o varias columnas en particular 
print(df[['user_id', 'time']].value_counts().sum())
##### ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
'''¿CÓMO PUEDE SER QUE UN USUARIO REPRODUZCA EN EL MISMO MOMENTO VARIAS CANCIONES?????
En teoría ya no hay filas exactas pero acabo de encontrar inconsistencias en los datos. 
Puede ser que solo es un proyecto de sprint hecho con ciclos y random values o puede ser que
el plan familiar de la plataforma pertenezca a un único user_id. Falta información para considerar la
inconsistencia'''

0
61253


'¿CÓMO PUEDE SER QUE UN USUARIO REPRODUZCA EN EL MISMO MOMENTO VARIAS CANCIONES?????\nEn teoría ya no hay filas exactas pero acabo de encontrar inconsistencias en los datos. \nPuede ser que solo es un proyecto de sprint hecho con ciclos y random values o puede ser que\nel plan familiar de la plataforma pertenezca a un único user_id. Falta información para considerar la\ninconsistencia'

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Las instrucciones solicitan dos métodos para calcular el número de duplicados explícitos, pero aquí solo se muestra uno. Considera añadir un segundo enfoque de conteo diferente (no solo repetir la misma expresión) para cumplir plenamente con el requisito.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [15]:
# Elimina los duplicados explícitos
df.drop_duplicates(inplace=True)



<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [23]:
# Comprueba de nuevo si hay filas exactamente duplicadas
print(df.duplicated().sum())
# Comprueba datos duplicados para dos columnas
print(df[['user_id', 'time']].value_counts().sum())

0
61253


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Requisito recurrente: esta sección también enfatiza demostrar dos maneras de contar duplicados. Considera mostrar aquí también el enfoque alternativo de conteo (aunque el resultado ahora sea 0).<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspecciona los nombres de géneros únicos
sorted(df['genre'].unique())

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [17]:
# Función para reemplazar los duplicados implícitos
'''
#La siguiente función, aunque es más eficiente, si algún valor no existe en la columna la función no hace nada
#Se necesita un error handling (no compete en esta etapa del curso)
def replace_wrong_values(df, column, wrong_values, correct_value):
    df[column].replace(wrong_values, correct_value, inplace=True)
'''
#Aunque más ineficiente contiene un bucle como lo pide el ejercicio.

def replace_wrong_values(df, column, wrong_values, correct_value):
    for i in wrong_values:
        df[column].replace(i, correct_value, inplace=True)
    

<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [18]:
# Elimina los duplicados implícitos
duplic_implicit_h =['hip', 'hop', 'hip-hop']

replace_wrong_values(df, 'genre', duplic_implicit_h, 'hiphop')


<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# LAs siguientes líneas solo imprimen los géneros que cominzan con la letra "h"
#jojo = [i for i in sorted(df['genre'].unique()) if i.startswith('h')]
#print(*jojo, sep='\n')
sorted(df['genre'].unique())

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Buen flujo de trabajo: inspeccionaste los géneros únicos, creaste una función reutilizable para corregir duplicados implícitos, la aplicaste para unificar las variantes de hiphop y luego verificaste la limpieza revisando de nuevo los valores únicos.<br>
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [20]:
# Obtiene el total de registros
print(f"El total de registros en el DataFrame es: {df.shape[0]}")

# Muestra la cantidad de usuarios totales para Shelbyville
shel_ids = df[df['city']=='Shelbyville']['user_id'].count()
# Muestra la cantidad de usuarios totales para Springfield
spri_ids = df[df['city']=='Springfield']['user_id'].count()
# Imprime la comparación entre los conteos de registros totales del dataframe con los de Shelbyville y Springfield
print(f"El total de registros en Shelbyville es: {shel_ids}\nEl total de registros en Springfield es: {spri_ids},\nEsto sumado coincide con los registros totales del DataFrame {shel_ids + spri_ids}.")
# Imprime Cuantas veces son mayores los registros en Springfield comparados con Shelbyville
print(f"La razón entre los registros totales de Springfield y de Shelbyvile es igual a {round(spri_ids/shel_ids, 2)}")
print()

# Imprime el número total de registros únicos en todo el DataFrame
print(f"El total de id únicos de usuario para ambas ciudades es: {df['user_id'].nunique()}.")
# Imprime el conteo de los usuarios únicos registrados en Shelviville
shel_unids = df[df['city']=='Shelbyville']['user_id'].nunique()
# Imprime el conteo de los usuarios únicos registrados en Springfield
spri_unids = df[df['city']=='Springfield']['user_id'].nunique()
# Imprime la comparación entre los conteos de registros únicos totales del dataframe con los registros únicos de Shelbyville y los registros únicos de Springfield
print(f"El total de registros únicos de usuario en Shelbyville es: {shel_unids}\nEl total de registros únicos de usuario en Springfield es: {spri_unids},\nEsto sumado coincide con los registros únicos totales del DataFrame {shel_ids + spri_ids}.")
# Imprime Cuantas veces son mayores los registros únicos en Springfield comparados con Shelbyville
print(f"La razón entre los registros por usuario (únicos) de Springfield y de Shelbyvile es igual a {round(spri_unids/shel_unids,2)}")
print()

# Cuenta las canciones reproducidas en cada ciudad
ciudad_cancion = df.groupby('city')['track'].count()
print(ciudad_cancion)


El total de registros en el DataFrame es: 61253
El total de registros en Shelbyville es: 18512
El total de registros en Springfield es: 42741,
Esto sumado coincide con los registros totales del DataFrame 61253.
La razón entre los registros totales de Springfield y de Shelbyvile es igual a 2.31

El total de id únicos de usuario para ambas ciudades es: 41748.
El total de registros únicos de usuario en Shelbyville es: 12423
El total de registros únicos de usuario en Springfield es: 29358,
Esto sumado coincide con los registros únicos totales del DataFrame 61253.
La razón entre los registros por usuario (únicos) de Springfield y de Shelbyvile es igual a 2.36

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Al dividir los registros de Springflied entre los de Shelbyville encontramos que los registros de Springfield son 2.3 veces más que los de shelbyvile.
Esto puede deberse al tamaño de la ciudad o que una ciudad tiene más usuarios registrados que la otra:

- Para probar si la diferencia se debe al tamaño de la ciudad, el DataFrame no nos da información suficiente; al realizar una búsqueda se encontró que existen muchas ciudades con esos nombres en diferentes estados de EUA. por lo que no se puede comprobar.
- Para probar si la diferencia se debe a la cantidad de usuarios para cada ciudad: Se realizaron unos filtros que muestran que la cantidad de usuarios únicos (cada usuario tiene un solo 'user_id' diferente al resto), los resultados arrojan que en Shelyville la cantidad de usuarios registrados por user id es 12,423 mientras que en Springfield 29,358.
    - Al dividir usuarios únicos de Springfield entre la cantidad de usuarios únicos en Shelbyville descubrimos que la cantidad de usuarios en springfield es 2.36 veces más grande que la cantidad de usuarios en Shelbyville.

Conclusión:

La similitud entre la razón de registros totales (2.3) y la razón de usuarios únicos (2.36) entre Springfield y Shelbyville sugiere que la diferencia en el número de registros se debe principalmente a la cantidad de usuarios registrados en cada ciudad y no a un comportamiento distinto dentro de la plataforma. Aunque Springfield cuenta con un mayor número de usuarios (posiblemente relacionado con el tamaño de la ciudad), el hecho de que ambas proporciones sean prácticamente iguales indica que los usuarios de ambas ciudades presentan patrones de uso similares. En otras palabras, cada usuario, en promedio, genera una cantidad comparable de registros en ambas ciudades, lo que permite inferir que el comportamiento de interacción con la plataforma es consistente entre Springfield y Shelbyville.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Interpretación clara: comparaste Springfield vs. Shelbyville con ratios concretos y conectaste la diferencia en el total de reproducciones con la diferencia en el número de usuarios únicos.<br>
</div>

Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [21]:
# Crea un nuevo dataframe con todas las columnas pero solo muestra las filas en las que la columna 'day' es lunes o viernes
dias_filtrados = df.query("day == 'Monday' or day == 'Friday'")
# Crea un nuevo dataframe con todas las columnas pero solo muestra las filas en las que la columna 'day' es lunes o viernes y la columna 'city' el Shelbyville
shel_dias_filtrados = df.query("(day == 'Monday' or day == 'Friday') and city =='Shelbyville'")
# Crea un nuevo dataframe con todas las columnas pero solo muestra las filas en las que la columna 'day' es lunes o viernes y la columna 'city' el Springfield
spri_dias_filtrados = df.query("(day == 'Monday' or day == 'Friday') and city =='Springfield'")

# Crea un series que muestra el conteo de las reproducciones para lunes y viernes de ambas ciudades
filtro_ambas = dias_filtrados.groupby('day')['track'].count()
# Crea un series que muestra el conteo de las reproducciones para lunes y viernes de la ciudad de Shelbyville
filtro_shel = shel_dias_filtrados.groupby('day')['track'].count()
# Crea un series que muestra el conteo de las reproducciones para lunes y viernes de la ciudad de Shelbyville
filtro_spri = spri_dias_filtrados.groupby('day')['track'].count()

# Imprime el series para ambas ciudades
print(f"Lunes y viernes para ambas ciudades:\n{filtro_ambas}\n")
# Imprime el series para Shelbyville
print(f"Lunes y viernes para Shelbyville:\n{filtro_shel}\n")
# Imprime el series para Springfield
print(f"Lunes y viernes para Springfield:\n{filtro_spri}\n")
print()

# Imprime la difenercia de reproducciones para ambos días y ambas ciudades
print(f"Para ambas ciudades existe una diferencia de {filtro_ambas[0]-filtro_ambas[1]} más reproducciones el viernes que el lunes")
print()
# Imprime la difenercia de reproducciones para ambos días en Shelbyville
print(f"Para la ciudad de Shelbyville existe una diferencia de {filtro_shel[0]-filtro_shel[1]} más reproducciones el viernes que el lunes")
print()
# Imprime la difenercia de reproducciones para ambos días en Springfield
print(f"Para la ciudad de Springfield existe una diferencia de {filtro_spri[0]-filtro_spri[1]} más reproducciones el viernes que el lunes")

Lunes y viernes para ambas ciudades:
day
Friday    21840
Monday    21354
Name: track, dtype: int64

Lunes y viernes para Shelbyville:
day
Friday    5895
Monday    5614
Name: track, dtype: int64

Lunes y viernes para Springfield:
day
Friday    15945
Monday    15740
Name: track, dtype: int64


Para ambas ciudades existe una diferencia de 486 más reproducciones el viernes que el lunes

Para la ciudad de Shelbyville existe una diferencia de 281 más reproducciones el viernes que el lunes

Para la ciudad de Springfield existe una diferencia de 205 más reproducciones el viernes que el lunes


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

El día con más actividad para ambas ciudades y por separado fue el día viernes teniendo un total de 486 más reproducciones tomando en cuenta ambas ciudades, 281 más reproducciones para la ciudad de Shelbyville y 205 más reproducciones para Springfield.

- En la conclusión pasada observamos que los comportamientos de los usuarios dentro de la plataforma para ambas cuidades son muy similares en general; sin embargo al ser Springfield una ciudad con más usuarios registrados que Shelbyville y viendo que Shelbyvile tiene mayor diferencia entre días que Springfield. Podríamos decir que los viernes tenemos más actividad en Shelbyville comparado a la actividad en Springfield.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Buen desglose por día y por ciudad: filtrar a lunes/viernes y resumir los resultados facilita comparar los patrones de actividad entre ambas ubicaciones.<br>
</div>

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [24]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
'''
Esto de abajo es la prueba que realicé para armar la función
filas_dia = df.query("day == 'Monday'")
print(filas_dia)
print()
filas_city = filas_dia.query("city == 'Shelbyville'")
print(filas_city)
filtro_listo=filas_city['user_id'].count()
print()
print(filtro_listo)
'''
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filtrado = df.query("day == @day and city == @city")
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    conteo = filtrado['user_id'].count()
    reg_por_usuario = filtrado['user_id'].nunique()
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    print(f"La cantidad de reproducciones de los usuarios el día {day} en {city} es {conteo}\ny la cantidad de usuarios que usaron la plataforma ese día es: {reg_por_usuario}")
    # regresa para datos posteriores el número pedido
    return conteo
    


<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

`number_tracks()` imprime un mensaje y devuelve el resultado de `print()` (que es `None`), pero la consigna pide que la función devuelva el conteo numérico para que pueda reutilizarse más adelante en el notebook. Considera devolver el valor del conteo desde la función y hacer cualquier impresión con formato fuera de la función.<br>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment #2</b> <a class="tocSkip"></a>

Perfecto, buen trabajo con la modificación. 
</div>

<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [25]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday', 'Springfield')

La cantidad de reproducciones de los usuarios el día Monday en Springfield es 15740
y la cantidad de usuarios que usaron la plataforma ese día es: 11904


15740

In [26]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday', 'Shelbyville')

La cantidad de reproducciones de los usuarios el día Monday en Shelbyville es 5614
y la cantidad de usuarios que usaron la plataforma ese día es: 4248


5614

In [27]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday', 'Springfield')

La cantidad de reproducciones de los usuarios el día Friday en Springfield es 15945
y la cantidad de usuarios que usaron la plataforma ese día es: 12145


15945

In [28]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday', 'Shelbyville')

La cantidad de reproducciones de los usuarios el día Friday en Shelbyville es 5895
y la cantidad de usuarios que usaron la plataforma ese día es: 4495


5895

<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


A lo largo del análisis se identificaron patrones claros en el comportamiento de los usuarios de la plataforma musical en las ciudades de Springfield y Shelbyville. En términos generales, los datos muestran que el comportamiento de los usuarios es muy similar entre ambas ciudades, ya que la razón entre el número total de registros y la cantidad de usuarios únicos resultó prácticamente equivalente. Esto indica que, aunque Springfield cuenta con más usuarios registrados, cada usuario interactúa con la plataforma de manera comparable en ambas ciudades.

Durante el proceso de limpieza y preprocesamiento se detectaron diversos problemas que podían afectar la calidad del análisis, como nombres de columnas mal escritos, uso inconsistente de formatos y la presencia de duplicados explícitos e implícitos en los géneros musicales. Estos problemas se solucionaron estandarizando los nombres de las columnas, eliminando registros duplicados y corrigiendo valores inconsistentes mediante el uso de estructuras de datos externas y funciones personalizadas. Estas acciones permitieron trabajar con datos más confiables y evitar conteos erróneos que hubieran distorsionado los resultados.

Al comparar la actividad por día de la semana, se observó que tanto en Springfield como en Shelbyville existe un incremento en el número de reproducciones los viernes en comparación con los lunes. Sin embargo, el aumento es mayor en Shelbyville, donde se registraron 281 reproducciones adicionales los viernes, frente a 205 en Springfield. Esto sugiere que, aunque el comportamiento general es similar, la ciudad de Shelbyville presenta una mayor intensidad de uso de la plataforma al final de la semana.

En conclusión, los datos no muestran diferencias significativas en el comportamiento básico de los usuarios entre ciudades, pero sí evidencian variaciones en la actividad según el día de la semana. La correcta limpieza y estandarización de los datos fue fundamental para llegar a estas conclusiones, ya que permitió realizar comparaciones más precisas y obtener resultados confiables que responden de manera clara a la pregunta central del proyecto.